crypto/rsa/rsa_lib.c

```diff
@@ -163,6 +163,124 @@ void *RSA_get_ex_data(const RSA *r, int idx)
     return CRYPTO_get_ex_data(&r->ex_data, idx);
 }
 
+/*
+ * Define a scaling constant for our fixed point arithmetic.
+ * This value must be a power of two because the base two logarithm code
+ * makes this assumption.  The exponent must also be a multiple of three so
+ * that the scale factor has an exact cube root.  Finally, the scale factor
+ * should not be so large that a multiplication of two scaled numbers
+ * overflows a 64 bit unsigned integer.
+ */
+static const unsigned int scale = 1 << 18;
+static const unsigned int cbrt_scale = 1 << (2 * 18 / 3);
+
+/* Define some constants, none exceed 32 bits */
+static const unsigned int log_2  = 0x02c5c8;    /* scale * log(2) */
+static const unsigned int log_e  = 0x05c551;    /* scale * log2(M_E) */
+static const unsigned int c1_923 = 0x07b126;    /* scale * 1.923 */
+static const unsigned int c4_690 = 0x12c28f;    /* scale * 4.690 */
+
+/*
+ * Multiply two scale integers together and rescale the result.
+ */
+static ossl_inline uint64_t mul2(uint64_t a, uint64_t b)
+{
+    /*
+     * If correct rounded is desired, enable this define.
+     * Correct rounding being:
+     *      1) if the fractional part is less than one half, round down.
+     *      2) if the fractional part is greater than one half, round up.
+     *      3) if the fractional part is exactly one half, round to even.
+     * This very slightly increases accuracy.
+     */
+#ifdef OSSL_CORRECT_ROUNDING
+    uint64_t m = a * b;
+    uint32_t f = m & (scale - 1);
+
+    m /= scale;
+    if (f == scale / 2)         /* Round to even */
+        m &= ~1;
+    else if (f >= scale / 2)    /* Round up */
+        m++;
+    return m;
+#else
+    return a * b / scale;
+#endif
+}
+
+/*
+ * Calculate the cube root of a 64 bit scaled integer.
+ * Although the cube root of a 64 bit number does fit into a 32 bit unsigned
+ * integer, this is guaranteed after scaling, so this function has a 64 bit
```

Comment:
**not** guaranteed

Reply:
maybe mention what this algorithm is - variant of the shifting nth root algorithm?